### PR TITLE
537 engineers teach physics course flow

### PR DESCRIPTION
**Pull request by goodviber**
### Context
Some courses are created as part of a campaign to improve recruitment. The Engineers Teach Physics course is part of such a campaign. It is ultimately just a physics course but it will need to have a different title as part of the campaign.
We'll need to build a new step in the course flow to allow providers to create these types of courses.
What needs to be done

Add new step in course creation flow to capture whether a course with a physics subject is an engineers teach physics course
Handle the ability to update the subject and re-collect the engineers teach physics information if the subject is being changed back to physics from confirmation page
Success

If physics subject is chosen, the engineers teach physics step is shown and a summary item for it is shown on the confirmation page
The engineers teach physics step can be updated to yes or no. If the subject is still physics, the summary for engineers teach physics should always be shown with a yes or no
If the subject is changed then the summary item for engineers teach physics should not be shown
If the subject is changed back to physics, the engineers teach physics information should be recollected
Tech notes

This is very similar to the visa work we did, so we should be able to build on top of that for this step
We'll need to make sure the title shows as Engineers Teach Physics for courses with a physics subject but then continue to follow the logic for additional subjects. Eg Engineers Teach Physics with Maths. The logic is here [publish-teacher-training/generate_course_name_service.rb at main · DFE-Digital/publish-teacher-training](https://github.com/DFE-Digital/publish-teacher-training/blob/main/app/services/courses/generate_course_name_service.rb)

### Changes proposed in this pull request
Engineers teach physics conditional step
### Guidance to review
There is pre-existing behaviour in the modern language step which alters the index of the subjects ids
### Checklist

- [ ] Make sure all information from the Trello card is in here
- [ ] Attach to Trello card
- [ ] Rebased main
- [ ] Cleaned commit history
- [ ] Tested by running locally

### CODE REVIEW

File: app/models/concerns/courses/edit_options/engineers_teach_physics_concern.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+module Courses
+  module EditOptions
+    module EngineersTeachPhysicsConcern
+      extend ActiveSupport::Concern
+      included do
+        # When changing anything here be sure to update the edit_options in the
+        # courses factory in publish-teacher-training:
+        #
+        # https://github.com/DFE-Digital/publish-teacher-training/blob/master/spec/factories/courses.rb
```
Comment (defong) [inline app/models/concerns/courses/edit_options/engineers_teach_physics_concern.rb:9]: Did you do it?

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -83,6 +83,7 @@
         </div>
       </main>
     </div>
+    <%= debug params %>
```
Comment (defong) [inline app/views/layouts/application.html.erb:86]: leftovers

File: app/views/publish/courses/engineers_teach_physics/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
+              <h1 class="govuk-fieldset__heading">
+                Engineers Teach Physics
+              </h1>
+            </legend>
+            <p class="govuk-body">
+              Engineers Teach Physics is a new campaign to help engineers become physics teachers.
+            </p>
+            <p class="govuk-body">
+              For more information, email <%= govuk_mail_to("itt.engagement@education.gov.uk", subject: "Engineers Teach Physics") %>
```
Comment (defong) [inline app/views/publish/courses/engineers_teach_physics/new.html.erb:33]: Can we stick the email in to settings please

File: db/schema.rb [hunk 4]
```diff
@@ -135,12 +135,14 @@
     t.boolean "can_sponsor_skilled_worker_visa", default: false
     t.boolean "can_sponsor_student_visa", default: false
     t.integer "campaign_name"
+    t.boolean "engineers_teach_physics_query", default: false
```
Comment (defong) [inline db/schema.rb:138]: I thought we was using `campaign_name`

File: app/controllers/publish/courses/applications_open_controller.rb [hunk 5]
```diff
@@ -31,6 +31,7 @@ def actual_params
             :subjects_ids,
             :goto_confirmation,
             :accredited_body_code,
+            :engineers_teach_physics_query,
```
Comment (defong) [inline app/controllers/publish/courses/applications_open_controller.rb:34]: can we drop `query`?

File: app/controllers/publish/courses/engineers_teach_physics_controller.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      end
+
+    private
+
+      def build_course_params
+        params[:course][:engineers_teach_physics_query] = params[:engineers_teach_physics_query]
+      end
+
+      def has_physics_subject?
+        @course.name.split.first == "Physics"
```
[the remark this answers is not in the document]
Comment (goodviber) [reply app/controllers/publish/courses/engineers_teach_physics_controller.rb]: It is a bit, tho ` params[:course].delete(:master_subject_id)` at line 82 ?

[the remark this answers is not in the document]
Comment (goodviber) [reply app/controllers/publish/courses/engineers_teach_physics_controller.rb]: Any reason why we can't keep the master_subject_id ?

File: app/models/concerns/courses/edit_options/engineers_teach_physics_concern.rb (hunk 1 above)
Comment (goodviber) [reply app/models/concerns/courses/edit_options/engineers_teach_physics_concern.rb:9]: I made a trait, but I can't see any `edit_options` in the courses factory?

File: app/views/layouts/application.html.erb (hunk 2 above)
Comment (goodviber) [reply app/views/layouts/application.html.erb:86]: Not finished yet

File: app/views/publish/courses/engineers_teach_physics/new.html.erb (hunk 3 above)
Comment (goodviber) [reply app/views/publish/courses/engineers_teach_physics/new.html.erb:33]: Good idea

File: app/models/concerns/courses/edit_options/engineers_teach_physics_concern.rb (hunk 1 above)
Comment (defong) [reply app/models/concerns/courses/edit_options/engineers_teach_physics_concern.rb:9]: Bin the comments?

File: app/controllers/publish/courses/engineers_teach_physics_controller.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (defong) [reply app/controllers/publish/courses/engineers_teach_physics_controller.rb]: It was ported along from the old `publish`

File: db/schema.rb (hunk 4 above)
Comment (goodviber) [reply db/schema.rb]: yes, updated

File: app/controllers/publish/courses/applications_open_controller.rb (hunk 5 above)
Comment (goodviber) [reply app/controllers/publish/courses/applications_open_controller.rb]: Yep, as above

File: app/models/concerns/courses/edit_options/engineers_teach_physics_concern.rb (hunk 1 above)
Comment (goodviber) [reply app/models/concerns/courses/edit_options/engineers_teach_physics_concern.rb]: Yep

File: app/controllers/publish/courses/engineers_teach_physics_controller.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+          redirect_to new_publish_provider_recruitment_cycle_courses_engineers_teach_physics_path(path_params)
+        else
+          redirect_to @back_link_path
+        end
+      end
+
+    private
+
+      def has_physics_subject?
+        @course.master_subject_id == 29
```
Comment (darokel) [inline app/controllers/publish/courses/engineers_teach_physics_controller.rb:30]: Can we rely on this? Would it be better to find the subject by its `subject_code` and then assign the id?

File: app/controllers/publish/courses/subjects_controller.rb [hunk 8]
```diff
@@ -79,7 +79,7 @@ def build_course_params
 
         params[:course][:subjects_ids] = selected_subject_ids
 
-        params[:course].delete(:master_subject_id)
+        # params[:course].delete(:master_subject_id)
```
Comment (darokel) [inline app/controllers/publish/courses/subjects_controller.rb:82]: Remnants?

File: app/models/course.rb [hunk 9]
```diff
@@ -48,7 +48,10 @@ class Course < ApplicationRecord
     not_required: 9,
   }
 
-  enum :campaign_name, COURSE_CAMPAIGNS
+  enum :campaign_name, {
+    no_campaign: "no_campaign",
```
Comment (darokel) [inline app/models/course.rb:52]: Do we need this? The presence of `nil` on this field should indicate it isn't a part of a campaign?

File: app/models/course.rb [hunk 10]
```diff
@@ -316,6 +319,8 @@ def self.entry_requirement_options_without_nil_choice
   validates :name, :profpost_flag, :program_type, :qualification, :start_date, :study_mode, presence: true
   validates :age_range_in_years, presence: true, on: %i[new create publish], unless: :further_education_course?
   validates :level, presence: true, on: %i[new create publish]
+  validates :campaign_name, inclusion: { in: campaign_names.keys }
+  # TODO: validates :master_subject_id ?
```
Comment (darokel) [inline app/models/course.rb:323]: Remnant?

File: app/views/layouts/application.html.erb (hunk 2 above)
Comment (darokel) [inline app/views/layouts/application.html.erb:86]: Remnant

File: app/views/publish/courses/engineers_teach_physics/_form_fields.html.erb [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    </legend>
+    <%= render "publish/shared/error_messages", error_keys: [:campaign_name] %>
+
+    <div class="govuk-radios govuk-!-margin-top-2" data-module="govuk-radios">
+
+      <% course.edit_course_options["engineers_teach_physics"].each do |value| %>
+        <div class="govuk-radios__item">
+          <%= form.radio_button :campaign_name,
+                                value,
+                                # checked: @course.campaign_name == value,
```
Comment (darokel) [inline app/views/publish/courses/engineers_teach_physics/_form_fields.html.erb:27]: Remnant?

File: config/locales/en.yml [hunk 12]
```diff
@@ -589,6 +591,11 @@ en:
         label: "Yes"
       false:
         label: "No"
+    engineers_teach_physics:
+      engineers_teach_physics:
```
Comment (darokel) [inline config/locales/en.yml:595]: Is this a typo (the double namespace)

File: db/migrate/20220928150349_add_campaign_name_to_course.rb [hunk 13]
```diff
@@ -1,5 +1,5 @@
 class AddCampaignNameToCourse < ActiveRecord::Migration[7.0]
   def change
-    add_column :course, :campaign_name, :integer
+    add_column :course, :campaign_name, :string, default: "no_campaign"
```
Comment (darokel) [inline db/migrate/20220928150349_add_campaign_name_to_course.rb:3]: Any rationale for this?

File: app/controllers/publish/courses/engineers_teach_physics_controller.rb (hunk 7 above)
Comment (goodviber) [reply app/controllers/publish/courses/engineers_teach_physics_controller.rb:30]: I am open to best suggestions here

File: app/controllers/publish/courses/subjects_controller.rb (hunk 8 above)
Comment (goodviber) [reply app/controllers/publish/courses/subjects_controller.rb:82]: Yes, see above

File: app/models/course.rb (hunk 9 above)
Comment (goodviber) [reply app/models/course.rb:52]: That is a bit problematic with the radio buttons and validations, we get into nil/false issues etc

File: app/models/course.rb (hunk 10 above)
Comment (goodviber) [reply app/models/course.rb:323]: Yes, see above, depends if we stick with `master_subject_id`

File: app/views/layouts/application.html.erb (hunk 2 above)
Comment (goodviber) [reply app/views/layouts/application.html.erb:86]: will rm when finished

File: db/migrate/20220928150349_add_campaign_name_to_course.rb (hunk 13 above)
Comment (goodviber) [reply db/migrate/20220928150349_add_campaign_name_to_course.rb:3]: Yep, the test specs would not seem to load the default from the initializer, played around and this seemed to work, also we should be able to set the default in the model, but that didn't work either...

File: app/controllers/publish/courses/engineers_teach_physics_controller.rb (hunk 7 above)
Comment (goodviber) [reply app/controllers/publish/courses/engineers_teach_physics_controller.rb:48]: But at the moment I think this is the best option

File: config/locales/en.yml (hunk 12 above)
Comment (goodviber) [reply config/locales/en.yml:595]: changed to campaign_name

File: app/controllers/publish/courses/subjects_controller.rb [hunk 14]
```diff
@@ -16,8 +16,19 @@ def continue
 
       def update
         authorize(provider)
+        if params[:course][:master_subject_id] == "29" && FeatureService.enabled?(:engineers_teach_physics_on_course)
```
Comment (darokel) [inline app/controllers/publish/courses/subjects_controller.rb:19]: I think this should use `SecondarySubject.physics.id`?

File: app/controllers/publish/courses/subjects_controller.rb [hunk 15]
```diff
@@ -16,8 +16,19 @@ def continue
...
+          course.update(master_subject_id: params[:course][:master_subject_id])
+          redirect_to(
+            engineers_teach_physics_publish_provider_recruitment_cycle_course_path(
+              @course.provider_code,
+              @course.recruitment_cycle_year,
+              @course.course_code,
+              course: { master_subject_id: "29", subjects_ids: selected_subject_ids },
```
[the remark this answers is not in the document]
Comment (goodviber) [reply app/controllers/publish/courses/subjects_controller.rb:26]: Yep

**Review by darokel — APPROVED** [review]
A couple of comments but snags have been noted and will be fixed in follow ups

File: app/controllers/publish/courses/subjects_controller.rb (hunk 14 above)
Comment (goodviber) [reply app/controllers/publish/courses/subjects_controller.rb:19]: Yep, to_s

File: spec/features/publish/courses/editing_engineers_teach_physics_spec.rb [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+  end
+
+  def modern_languages_subject_ids
+    subordinate_subject = course_subject(:modern_languages)
+    course_subject = course_subject(:physics)
+    "course%5Bsubjects_ids%5D%5B%5D=#{course_subject.id}&course%5Bsubjects_ids%5D%5B%5D=#{subordinate_subject.id}"
+  end
+end
+
+# /publish/organisations/A02/2023/courses/C251/engineers_teach_physics?course%5Bmaster_subject_id%5D=29&course%5Bsubjects_ids%5D%5B%5D=29&course%5Bsubjects_ids%5D%5B%5D=33
```
[the remark this answers is not in the document]
Comment (goodviber) [reply spec/features/publish/courses/editing_engineers_teach_physics_spec.rb:129]: Yep

**Pull request merged**